packages/remote/src/electron-node/setup/remote-copy-service.ts

```diff
@@ -0,0 +1,114 @@
+import { RemoteNativeDependencyService } from './remote-native-dependency-service';
+import { ContributionProvider } from '@theia/core';
+import * as archiver from 'archiver';
+import * as path from 'path';
+import * as fs from 'fs';
+import * as os from 'os';
+
+@injectable()
+export class RemoteCopyService {
+
+    @inject(ApplicationPackage)
+    protected readonly applicationPackage: ApplicationPackage;
+
+    @inject(RemoteCopyRegistry)
+    protected readonly copyRegistry: RemoteCopyRegistry;
+
+    @inject(RemoteNativeDependencyService)
+    protected readonly nativeDependencyService: RemoteNativeDependencyService;
+
+    @inject(ContributionProvider) @named(RemoteCopyContribution)
+    protected readonly copyContributions: ContributionProvider<RemoteCopyContribution>;
+
+    protected initialized = false;
+
+    async copyToRemote(remote: RemoteConnection, remotePlatform: RemotePlatform, destination: string): Promise<void> {
+        const zipName = path.basename(destination);
+        const projectPath = this.applicationPackage.projectPath;
+        const tempDir = await this.getTempDir();
+        const zipPath = path.join(tempDir, zipName);
+        const files = await this.getFiles(remotePlatform, tempDir);
+        // We stream to a file here and then copy it because it is faster
+        // Copying files via sftp is 4x times faster compared to readable streams
+        const stream = fs.createWriteStream(zipPath);
+        const archive = archiver('tar', {
+            gzip: true
+        });
+        archive.pipe(stream);
+        for (const file of files) {
+            const filePath = path.isAbsolute(file.path)
+                ? file.path
+                : path.join(projectPath, file.path);
+
+            archive.file(filePath, {
+                name: file.target,
+                mode: file.options?.mode
+            });
+        }
+        await archive.finalize();
+        await remote.copy(zipPath, destination);
+        await fs.promises.rm(tempDir, {
+            recursive: true,
+            force: true
+        });
+    }
+
+    protected async getFiles(remotePlatform: RemotePlatform, tempDir: string): Promise<RemoteFile[]> {
+        const [localFiles, nativeDependencies] = await Promise.all([
+            this.loadCopyContributions(),
+            this.loadNativeDependencies(remotePlatform, tempDir)
+        ]);
+        return [...localFiles, ...nativeDependencies];
+    }
+
+    protected async loadCopyContributions(): Promise<RemoteFile[]> {
+        if (this.initialized) {
+            return this.copyRegistry.getFiles();
+        }
+        await Promise.all(this.copyContributions.getContributions()
+            .map(copyContribution => copyContribution.copy(this.copyRegistry)));
+        this.initialized = true;
+        return this.copyRegistry.getFiles();
+    }
+
+    protected async loadNativeDependencies(remotePlatform: RemotePlatform, tempDir: string): Promise<RemoteFile[]> {
```

Comment:
So we're downloading the native dependencies to the local machine and add it to the zip to be sent. I guess that makes sense, since we're not sure the remote has access via http, right?

Reply:
Exactly, neither http ingress nor egress is garanteed.